File: 0173-Binary-Search-Tree-Iterator/0173.java
class BSTIterator {
  private TreeNode cur;
  private Deque<TreeNode> stack;

  public BSTIterator(TreeNode root) {
    cur = root;
    stack = new LinkedList<TreeNode>();
  }

  public int next() {
    while (cur != null) {
      stack.push(cur);
      cur = cur.left;
    }
    cur = stack.pop();
    int ret = cur.val;
    cur = cur.right;
    return ret;
  }

  public boolean hasNext() {
    return cur != null || !stack.isEmpty();
  }
}

File: 101-200/0173-Binary-Search-Tree-Iterator/0173.go
type BSTIterator struct {
	stack []*TreeNode
	cur   *TreeNode
}

func Constructor(root *TreeNode) BSTIterator {
	return BSTIterator{cur: root}
}

func (it *BSTIterator) Next() int {
	for node := it.cur; node != nil; node = node.Left {
		it.stack = append(it.stack, node)
	}
	it.cur, it.stack = it.stack[len(it.stack)-1], it.stack[:len(it.stack)-1]
	val := it.cur.Val
	it.cur = it.cur.Right
	return val
}

func (it *BSTIterator) HasNext() bool {
	return it.cur != nil || len(it.stack) > 0
}